Verify storage unit is consistent with the current proxy protocol when REGISTER STORAGE UNITS;

master [f52f4846](https://github.com/apache/shardingsphere/commit/f52f4846ce4cdce734dd79dcad1f94e1c50d408d)

`server.yaml`

```
mode:
  type: Cluster
  repository:
    type: ZooKeeper
    props:
      namespace: governance_ds
      server-lists: localhost:2181
      retryIntervalMilliseconds: 500
      timeToLiveSeconds: 60
      maxRetries: 3
      operationTimeoutMilliseconds: 500

authority:
  users:
    - user: root@%
      password: root
    - user: sharding
      password: sharding
  privilege:
    type: ALL_PERMITTED

props:
  proxy-frontend-database-protocol-type: PostgreSQL
```

`DistSQL`

```
REGISTER STORAGE UNIT ds_2 (
    URL="jdbc:mysql://127.0.0.1:3306/migration_ds_10",
    USER="root",
    PASSWORD="123456",
    PROPERTIES("minPoolSize"="1","maxPoolSize"="20","idleTimeout"="60000")
), ds_3 (
    URL="jdbc:mysql://127.0.0.1:3306/migration_ds_11",
    USER="root",
    PASSWORD="123456",
    PROPERTIES("minPoolSize"="1","maxPoolSize"="20","idleTimeout"="60000")
), ds_4 (
    URL="jdbc:mysql://127.0.0.1:3306/migration_ds_12",
    USER="root",
    PASSWORD="123456",
    PROPERTIES("minPoolSize"="1","maxPoolSize"="20","idleTimeout"="60000")
);
```

```
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unknown database 'pg_catalog'
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:944)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3978)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3914)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2530)
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2683)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2491)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2449)
	at com.mysql.jdbc.StatementImpl.executeQuery(StatementImpl.java:1381)
	at com.zaxxer.hikari.pool.ProxyStatement.executeQuery(ProxyStatement.java:110)
	at com.zaxxer.hikari.pool.HikariProxyStatement.executeQuery(HikariProxyStatement.java)
	at org.apache.shardingsphere.infra.metadata.data.collector.ShardingSphereTableDataCollectorUtil.collectRowData(ShardingSphereTableDataCollectorUtil.java:56)
	at org.apache.shardingsphere.infra.metadata.data.collector.tables.PgNamespaceTableCollector.collect(PgNamespaceTableCollector.java:48)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForTable(ShardingSphereDataScheduleCollector.java:111)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$collectForSchema$2(ShardingSphereDataScheduleCollector.java:98)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForSchema(ShardingSphereDataScheduleCollector.java:96)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$collectForDatabase$1(ShardingSphereDataScheduleCollector.java:89)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForDatabase(ShardingSphereDataScheduleCollector.java:87)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$run$0(ShardingSphereDataScheduleCollector.java:79)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.run(ShardingSphereDataScheduleCollector.java:77)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
[ERROR] 2023-01-18 13:42:09.959 [ShardingSphere-data-collect-0] o.a.s.d.p.c.e.ShardingSphereDataScheduleCollector - Collect data failed!
com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: Unknown database 'pg_catalog'
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:944)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3978)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3914)
	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2530)
	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2683)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2491)
	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2449)
	at com.mysql.jdbc.StatementImpl.executeQuery(StatementImpl.java:1381)
	at com.zaxxer.hikari.pool.ProxyStatement.executeQuery(ProxyStatement.java:110)
	at com.zaxxer.hikari.pool.HikariProxyStatement.executeQuery(HikariProxyStatement.java)
	at org.apache.shardingsphere.infra.metadata.data.collector.ShardingSphereTableDataCollectorUtil.collectRowData(ShardingSphereTableDataCollectorUtil.java:56)
	at org.apache.shardingsphere.infra.metadata.data.collector.tables.PgClassTableCollector.collect(PgClassTableCollector.java:54)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForTable(ShardingSphereDataScheduleCollector.java:111)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$collectForSchema$2(ShardingSphereDataScheduleCollector.java:98)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForSchema(ShardingSphereDataScheduleCollector.java:96)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$collectForDatabase$1(ShardingSphereDataScheduleCollector.java:89)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.collectForDatabase(ShardingSphereDataScheduleCollector.java:87)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.lambda$run$0(ShardingSphereDataScheduleCollector.java:79)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.apache.shardingsphere.data.pipeline.core.execute.ShardingSphereDataScheduleCollector$ShardingSphereDataCollectorRunnable.run(ShardingSphereDataScheduleCollector.java:77)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/metadata/data/ShardingStatisticsTableCollector.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/metadata/data/ShardingStatisticsTableCollector.java
@@ -87,13 +87,15 @@ public final class ShardingStatisticsTableCollector implements ShardingSphereDat
                 row.add(each.getLogicTable());
                 row.add(dataNode.getDataSourceName());
                 row.add(dataNode.getTableName());
-                addTableRowsAndDataLength(shardingSphereDatabase.getResourceMetaData().getDataSources(), dataNode, row, shardingSphereDatabase.getProtocolType());
+                addTableRowsAndDataLength(shardingSphereDatabase.getResourceMetaData().getStorageTypes(), shardingSphereDatabase.getResourceMetaData().getDataSources(), dataNode, row);
                 tableData.getRows().add(new ShardingSphereRowData(row));
             }
         }
     }
     
-    private void addTableRowsAndDataLength(final Map<String, DataSource> dataSources, final DataNode dataNode, final List<Object> row, final DatabaseType databaseType) throws SQLException {
+    private void addTableRowsAndDataLength(final Map<String, DatabaseType> databaseTypes, final Map<String, DataSource> dataSources,
+                                           final DataNode dataNode, final List<Object> row) throws SQLException {
+        DatabaseType databaseType = databaseTypes.get(dataNode.getDataSourceName());
         if (databaseType instanceof MySQLDatabaseType) {
             addForMySQL(dataSources, dataNode, row);
         } else if (databaseType instanceof PostgreSQLDatabaseType) {

--- a/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/ShardingSphereTableDataCollectorUtil.java
+++ b/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/ShardingSphereTableDataCollectorUtil.java
@@ -18,6 +18,7 @@
 package org.apache.shardingsphere.infra.metadata.data.collector;
 
 import org.apache.shardingsphere.infra.metadata.data.ShardingSphereRowData;
+import org.apache.shardingsphere.infra.metadata.database.ShardingSphereDatabase;
 import org.apache.shardingsphere.infra.metadata.database.schema.decorator.model.ShardingSphereColumn;
 import org.apache.shardingsphere.infra.metadata.database.schema.decorator.model.ShardingSphereTable;
 
@@ -28,6 +29,7 @@ import java.sql.SQLException;
 import java.sql.Statement;
 import java.sql.Types;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
 
@@ -39,17 +41,20 @@ public final class ShardingSphereTableDataCollectorUtil {
     /**
      * Collect row data.
      *
-     * @param dataSources data sources
+     * @param shardingSphereDatabase ShardingSphere database
      * @param sql sql
      * @param table table
      * @param selectedColumnNames selected column names
      * @return ShardingSphere row data
      * @throws SQLException sql exception
      */
-    public static Collection<ShardingSphereRowData> collectRowData(final Collection<DataSource> dataSources, final String sql, final ShardingSphereTable table,
+    public static Collection<ShardingSphereRowData> collectRowData(final ShardingSphereDatabase shardingSphereDatabase, final String sql, final ShardingSphereTable table,
                                                                    final Collection<String> selectedColumnNames) throws SQLException {
+        if (isProtocolTypeAndStorageTypeDifferent(shardingSphereDatabase)) {
+            return Collections.emptyList();
+        }
         Collection<ShardingSphereRowData> result = new LinkedList<>();
-        for (DataSource each : dataSources) {
+        for (DataSource each : shardingSphereDatabase.getResourceMetaData().getDataSources().values()) {
             try (
                     Connection connection = each.getConnection();
                     Statement statement = connection.createStatement();
@@ -58,6 +63,10 @@ public final class ShardingSphereTableDataCollectorUtil {
             }
         }
         return result;
+    }
+    
+    private static boolean isProtocolTypeAndStorageTypeDifferent(final ShardingSphereDatabase shardingSphereDatabase) {
+        return !shardingSphereDatabase.getResourceMetaData().getStorageTypes().values().stream().allMatch(each -> each.getType().equals(shardingSphereDatabase.getProtocolType().getType()));
     }
     
     private static Collection<ShardingSphereRowData> getRows(final ResultSet resultSet, final ShardingSphereTable table, final Collection<String> selectedColumnNames) throws SQLException {

--- a/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/tables/PgClassTableCollector.java
+++ b/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/tables/PgClassTableCollector.java
@@ -51,7 +51,7 @@ public final class PgClassTableCollector implements ShardingSphereDataCollector 
     @Override
     public Optional<ShardingSphereTableData> collect(final String databaseName, final ShardingSphereTable table,
                                                      final Map<String, ShardingSphereDatabase> shardingSphereDatabases) throws SQLException {
-        Collection<ShardingSphereRowData> rows = ShardingSphereTableDataCollectorUtil.collectRowData(shardingSphereDatabases.get(databaseName).getResourceMetaData().getDataSources().values(),
+        Collection<ShardingSphereRowData> rows = ShardingSphereTableDataCollectorUtil.collectRowData(shardingSphereDatabases.get(databaseName),
                 SELECT_SQL, table, Arrays.stream(COLUMN_NAMES.split(",")).map(String::trim).collect(Collectors.toList()));
         Collection<ShardingSphereRowData> rowData = decorateTableName(rows, table, shardingSphereDatabases.get(databaseName).getRuleMetaData().getRules());
         ShardingSphereTableData result = new ShardingSphereTableData(PG_CLASS);

--- a/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/tables/PgNamespaceTableCollector.java
+++ b/infra/common/src/main/java/org/apache/shardingsphere/infra/metadata/data/collector/tables/PgNamespaceTableCollector.java
@@ -45,7 +45,7 @@ public final class PgNamespaceTableCollector implements ShardingSphereDataCollec
     @Override
     public Optional<ShardingSphereTableData> collect(final String databaseName, final ShardingSphereTable table,
                                                      final Map<String, ShardingSphereDatabase> shardingSphereDatabases) throws SQLException {
-        Collection<ShardingSphereRowData> rows = ShardingSphereTableDataCollectorUtil.collectRowData(shardingSphereDatabases.get(databaseName).getResourceMetaData().getDataSources().values(),
+        Collection<ShardingSphereRowData> rows = ShardingSphereTableDataCollectorUtil.collectRowData(shardingSphereDatabases.get(databaseName),
                 SELECT_SQL, table, Arrays.stream(COLUMN_NAMES.split(",")).map(String::trim).collect(Collectors.toList()));
         ShardingSphereTableData result = new ShardingSphereTableData(PG_NAMESPACE);
         result.getRows().addAll(rows);

--- a/mode/core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java
+++ b/mode/core/src/main/java/org/apache/shardingsphere/mode/manager/ContextManager.java
@@ -274,6 +274,8 @@ public final class ContextManager implements AutoCloseable {
             MetaDataContexts reloadMetaDataContexts = createMetaDataContexts(databaseName, switchingResource, null);
             reloadMetaDataContexts.getMetaData().getDatabase(databaseName).getSchemas().forEach((schemaName, schema) -> reloadMetaDataContexts.getPersistService().getDatabaseMetaDataService()
                     .persist(reloadMetaDataContexts.getMetaData().getActualDatabaseName(databaseName), schemaName, schema));
+            Optional.ofNullable(reloadMetaDataContexts.getShardingSphereData().getDatabaseData().get(databaseName)).ifPresent(optional -> optional.getSchemaData().forEach((schemaName, schemaData) ->
+                    reloadMetaDataContexts.getPersistService().getShardingSphereDataPersistService().persist(databaseName, schemaName, schemaData, metaDataContexts.getMetaData().getDatabases())));
             alterSchemaMetaData(databaseName, reloadMetaDataContexts.getMetaData().getDatabase(databaseName), metaDataContexts.getMetaData().getDatabase(databaseName));
             metaDataContexts = reloadMetaDataContexts;
             metaDataContexts.getMetaData().getDatabases().putAll(newShardingSphereDatabase(metaDataContexts.getMetaData().getDatabase(databaseName)));

